### DISCUSSION TITLE
Corruption of strings

### DISCUSSION THREAD

**Discussion by earthboundkid** [progrium/darwinkit#269] [Q&A]
I'm seeing an issue where a Go string is sometimes corrupted when passed to darwinkit. Is there some magic thing I should do first to ensure the string isn't overwritten? I don't understand how they are supposed to interact. I tried just doing a defense strings.Clone, but somehow it didn't help and I still got corrupted memory.

**Comment by earthboundkid** [comment]
This is the code snippet:

```go
func buildName(path string) (string, error) {
	ppath := strings.Clone(path)
	// fmt.Println("*", path, ppath) // ← if I uncomment this, I don't get corruption!!
	dateAdded, err := getDateAdded(ppath)
	fmt.Println("*", path, ppath)
	if err != nil {
		return "", err
	}
	kind := getKind(path)
	name := filepath.Base(path)
	return fmt.Sprintf("%d/%02d/%s/%s", dateAdded.Year(), dateAdded.Month(), kind, name), nil
}

func getDateAdded(path string) (time.Time, error) {
	var dateAdded foundation.Date
	defer dateAdded.Release()
	var err foundation.Error
	defer err.Release()

	s := strings.Clone(path) // Apparently, Foundation breaks strings sometimes!!
	url := foundation.URL_FileURLWithPath(s)
	defer url.Release()
	if ok := url.GetResourceValueForKeyError(
		unsafe.Pointer(&dateAdded),
		foundation.URLAddedToDirectoryDateKey,
		unsafe.Pointer(&err),
	); !ok {
		return time.Time{}, fmt.Errorf("could not read %q", path)
	}

	unixTimestamp := dateAdded.TimeIntervalSince1970()
	seconds := math.Floor(float64(unixTimestamp))
	nanoseconds := (float64(unixTimestamp) - seconds) * 1e9

	return time.Unix(int64(seconds), int64(nanoseconds)), nil
}
```

**Reply by earthboundkid** [reply]
I'm doing some debugging, and it seems that the reason the defensive clone doesn't work is that \x00 is getting written on top of an unrelated string before the clone happens. This seems kind of bad: Foundation is just spewing nulls into random string space that it does not own. :-/

**Reply by earthboundkid** [reply]
I tried shoving everything into an autorelease pool and I'm not seeing the same corruption, although who knows if I just pushed it somewhere else: https://github.com/earthboundkid/scooter/commit/099fe29bcd3d772b331f8b7975be08b4745c2ac8#diff-bdce60bc059a47490191bb05e74205143a001e7cd0b81c358f0a77e468ea8a6fR168

**Reply by progrium** [reply]
Sorry for delay. But yeah, if you aren't using an app framework and just using some Foundation APIs directly you'll have to manually use autorelease pools. Though, all that does is ensure any allocations with Autorelease called (anything returned by a New* constructor in Go) are released. So I'm not sure what's happening with the string, I haven't seen this before. Maybe @programmingkidx has?

**Reply by earthboundkid** [reply]
I think part of what was going on was whatever got the string expected it to be null terminated. When I changed to passing the string with a null just tacked onto the end, it still worked, which suggests something down the line was expecting a null terminated string:

```go
var (
	ok            bool
	unixTimestamp float64
)
// It can't hurt to add a null terminator just to be safe
s := strings.Clone(path + "\x00")

// Was getting random memory corruption,
// so let's try just throwing in a pool
objc.WithAutoreleasePool(func() {
	var dateAdded foundation.Date
	var err foundation.Error
	url := foundation.NewURLFileURLWithPath(s)
	ok = url.GetResourceValueForKeyError(
		unsafe.Pointer(&dateAdded),
		foundation.URLAddedToDirectoryDateKey,
		unsafe.Pointer(&err),
	)
	if !ok {
		return
	}
	unixTimestamp = float64(dateAdded.TimeIntervalSince1970())
})
```

**Reply by programmingkidx** [reply]
> Maybe @programmingkidx has?

Sorry I haven't seen this issue.

**Reply by torarnv** [reply]
I'm seeing a very similar issue, where data structures in my app are corrupted as a result of calling APIs via darwinkit. I have a list of objects, each with a rune, now and then one of the runes become null/`\0`. If I remove/disable the calls via darwinkit, the problem seemingly goes away. The APIs I'm calling are not passed my data structure. Very strange.

**Reply by torarnv** [reply]
And I do have an `objc.WithAutoreleasePool` around my code

**Reply by earthboundkid** [reply]
I also saw nulls pop up on a seemingly random other object. Something must be "intelligently" inserting nulls into strings to convert them from length plus pointer to null terminated and end up just spraying the nulls randomly. :-/

**Reply by progrium** [reply]
Share any specific examples to reproduce. I haven't run into this yet but hopefully somebody will have the time to investigate.
